### PR TITLE
[RayJob] Add JobDeploymentStatusFailed Status and Reason Field to Enhance Observability for Flyte/RayJob Integration

### PR DESCRIPTION
**Pull request by Yicheng-Lu-llll**
## Why are these changes needed?

After this PR:

1. There will be two terminal deployment statuses(Inspired by [how K8s jobs define the status](https://sourcegraph.com/github.com/kubernetes/kubernetes@fc7325fbec0fe379694e58b82e09f99f166c5e6d/-/blob/staging/src/k8s.io/api/batch/v1/types.go?L531-533) and [how the Spark operator defines the terminal statuses](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/v1beta2-1.3.8-3.1.1/pkg/controller/sparkapplication/controller.go#L615)): 
    - `JobDeploymentStatusComplete`
    - `JobDeploymentStatusFailed`.  

2. If RayJob runs into `JobDeploymentStatusFailed`(Inspired by [how k8s job use `reason` and `message` fields](https://sourcegraph.com/github.com/kubernetes/kubernetes@fc7325fbec0fe379694e58b82e09f99f166c5e6d/-/blob/staging/src/k8s.io/api/batch/v1/types.go?L569-574)):
    - the `reason` field will include a machine-readable string that describes the failure reason. Other controllers like Flyte, can take actions based on this field.
    - The `message` field will include a human-readable explanation of the failure reason.

3. There are three failed reason currently:
    - `SubmissionFailed`: The submitter job failed to submit the user code.
    - `DeadlineExceeded`: This is analogous to `JobReasonDeadlineExceeded` in Kubernetes Jobs.
    - `AppFailed`: The user code throw an error.

### Additional Note:
Ideally, users should only consider the `reason` and `message` fields when the Deployment Status is `Failed`. 
But Considering the fact that the submitter job might retry, these two fields have slightly different behavior：

- The `reason` field is set only when the Deployment Status changes to `Failed`. Therefore, no failure reason will be displayed during the retry period, as the Deployment Status remains `running`. 
- the `message` field will always display the message for the current run. So, might have error message even if the Deployment Status remains `running`.

## Checks
I have changed the CI test accordingly; all changes are covered by the CI test, but I have also manually tested it to give an idea of how the current status looks like:

- The submitter job failed to submit the user code:
```
  Job Deployment Status:  Failed
  Job Id:                 rayjob-sample-wd89p
  Message:                Job submission has failed. Reason: BackoffLimitExceeded. Message: Job has reached the specified backoff limit
  Reason:      SubmissionFailed
```

- deadline exceed:
```
  End Time:               2024-02-24T21:58:47Z
  Job Deployment Status:  Failed
  Job Id:                 rayjob-sample-7gj87
  Message:                The RayJob has passed the activeDeadlineSeconds. StartTime: 2024-02-24 21:58:46 +0000 UTC. ActiveDeadlineSeconds: 1
  Reason:      DeadlineExceeded
```
- The user code throw an error:
```
  Job Deployment Status:  Failed
  Job Id:                 rayjob-sample-5bk57
  Job Status:             FAILED
  Message:                Job entrypoint command failed with exit code 1, last available logs (truncated to 20,000 chars):
2024-02-24 14:02:23,886   INFO worker.py:1715 -- Connected to Ray cluster. View the dashboard at ^[[1m^[[32mhttp://10.244.0.13:8265 ^[[39m^[[22m
test_counter got 1
test_counter got 2
test_counter got 3
test_counter got 4
test_counter got 5
Traceback (most recent call last):
  File "/home/ray/samples/sample_code.py", line 28, in <module>
    assert requests.__version__ == "2.26.0dewfrgsthsnrdyukjuytretyhgjk"
AssertionError
  Reason:                 AppFailed
```
- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 1]
```diff
@@ -674,20 +677,32 @@ func (r *RayJobReconciler) updateStatusToSuspendingIfNeeded(ctx context.Context,
 func (r *RayJobReconciler) checkK8sJobAndUpdateStatusIfNeeded(ctx context.Context, rayJob *rayv1.RayJob, job *batchv1.Job) bool {
 	for _, cond := range job.Status.Conditions {
 		if cond.Type == batchv1.JobFailed && cond.Status == corev1.ConditionTrue {
-			r.Log.Info("The submitter Kubernetes Job has failed. Attempting to transition the status to `Complete`.", "RayJob", rayJob.Name, "Submitter K8s Job", job.Name, "Reason", cond.Reason, "Message", cond.Message)
-			rayJob.Status.Message = "The submitter Kubernetes Job is failed. Reason: " + cond.Reason + ". Message: " + cond.Message
-			rayJob.Status.JobDeploymentStatus = rayv1.JobDeploymentStatusComplete
+			r.Log.Info("The submitter Kubernetes Job has failed. Attempting to transition the status to `Failed`.", "RayJob", rayJob.Name, "Submitter K8s Job", job.Name, "Reason", cond.Reason, "Message", cond.Message)
+			rayJob.Status.JobDeploymentStatus = rayv1.JobDeploymentStatusFailed
+			// The Submitter Job needs to wait for the user code to finish and retrieve its logs.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:682]:
```suggestion
			// The submitter Job needs to wait for the user code to finish and retrieve its logs.
```

File: ray-operator/apis/ray/v1/rayjob_types.go [hunk 2]
```diff
@@ -38,10 +38,17 @@ const (
 	JobDeploymentStatusInitializing JobDeploymentStatus = "Initializing"
 	JobDeploymentStatusRunning      JobDeploymentStatus = "Running"
 	JobDeploymentStatusComplete     JobDeploymentStatus = "Complete"
+	JobDeploymentStatusFailed       JobDeploymentStatus = "Failed"
 	JobDeploymentStatusSuspending   JobDeploymentStatus = "Suspending"
 	JobDeploymentStatusSuspended    JobDeploymentStatus = "Suspended"
 )
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1/rayjob_types.go:45]: Define a new type `type JobFailedReason string`?

File: ray-operator/apis/ray/v1/rayjob_types.go [hunk 3]
```diff
@@ -38,10 +38,17 @@ const (
...
 
+const (
+	JobReasonSubmissionFailed       string = "SubmissionFailed"
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1/rayjob_types.go:47]:
```suggestion
	SubmissionFailed       string = "SubmissionFailed"
```

File: ray-operator/apis/ray/v1/rayjob_types.go [hunk 4]
```diff
@@ -38,10 +38,17 @@ const (
...
+	JobReasonDeadlineExceeded       string = "DeadlineExceeded"
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1/rayjob_types.go:48]:
```suggestion
	DeadlineExceeded       string = "DeadlineExceeded"
```

File: ray-operator/apis/ray/v1/rayjob_types.go [hunk 5]
```diff
@@ -38,10 +38,17 @@ const (
...
+	JobReasonApplicationLevelFailed string = "ApplicationLevelFailed"
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1/rayjob_types.go:49]:
```suggestion
	AppFailed string = "AppFailed"
```

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 6]
```diff
@@ -180,19 +178,17 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
...
 		job := &batchv1.Job{}
 		if rayJobInstance.Spec.SubmissionMode == rayv1.K8sJobMode {
-			// If the submitting Kubernetes Job reaches the backoff limit, transition the status to `Complete`. This is because,
-			// beyond this point, it becomes impossible for the submitter to submit any further Ray jobs. For light-weight mode,
-			// we don't transition the status to `Complete` based on the number of failed requests. Instead, users can use the
-			// `ActiveDeadlineSeconds` to ensure that the RayJob in the light-weight mode is not stuck in the `Running` status
-			// indefinitely.
+			// If the submitting Kubernetes Job reaches the backoff limit, transition the status to `Complete` or `Failed`.
+			// This is because, beyond this point, it becomes impossible for the submitter to submit any further Ray jobs.
+			// For light-weight mode, we don't transition the status to `Complete` or Failed` based on the number of failed requests.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:189]:
```suggestion
			// For light-weight mode, we don't transition the status to `Complete` or `Failed` based on the number of failed requests.
```

File: ray-operator/apis/ray/v1/rayjob_types.go (hunk 5 above)
Comment (andrewsykim) [reply ray-operator/apis/ray/v1/rayjob_types.go:49]: JobFailed?

Comment (Yicheng-Lu-llll) [reply ray-operator/apis/ray/v1/rayjob_types.go:49]: I feel users might be confused about whether it is a RayJob (KubeRay), a job (Ray), or a Kubernetes submitter job if we use 'JobFailed'.

**Review by kevin85421 — APPROVED** [review]
LGTM

**Comment by kevin85421** [comment]
cc @pingsutw

**Comment by pingsutw** [comment]
Thank you @kevin85421 @Yicheng-Lu-llll

**Pull request merged**